llama_index/prompts/guidance_utils.py

```diff
@@ -0,0 +1,156 @@
+from typing import TYPE_CHECKING, Optional, Type, TypeVar
+
+from llama_index.output_parsers.base import OutputParserException
+from llama_index.output_parsers.utils import parse_json_markdown
+
+if TYPE_CHECKING:
+    from guidance import Program
+
+from pydantic import BaseModel
+
+
+def convert_to_handlebars(text: str) -> str:
+    """Convert a python format string to handlebars-style template.
+
+    In python format string, single braces {} are used for variable substitution,
+        and double braces {{}} are used for escaping actual braces (e.g. for JSON dict)
+    In handlebars template, double braces {{}} are used for variable substitution,
+        and single braces are actual braces (e.g. for JSON dict)
+
+    This is currently only used to convert a python format string based prompt template
+    to a guidance program template.
+    """
+    # Replace double braces with a temporary placeholder
+    var_left = "TEMP_BRACE_LEFT"
+    var_right = "TEMP_BRACE_RIGHT"
+    text = text.replace("{{", var_left)
+    text = text.replace("}}", var_right)
+
+    # Replace single braces with double braces
+    text = text.replace("{", "{{")
+    text = text.replace("}", "}}")
+
+    # Replace the temporary placeholder with single braces
+    text = text.replace(var_left, "{")
+    text = text.replace(var_right, "}")
+    return text
+
+
+def wrap_json_markdown(text: str) -> str:
+    """Wrap text in json markdown formatting block."""
+    return "```json\n" + text + "\n```"
+
+
+def pydantic_to_guidance_output_template(cls: Type[BaseModel]) -> str:
+    """Convert a pydantic model to guidance output template."""
+    return json_schema_to_guidance_output_template(cls.schema(), root=cls.schema())
+
+
+def pydantic_to_guidance_output_template_markdown(cls: Type[BaseModel]) -> str:
+    """Convert a pydantic model to guidance output template wrapped in json markdown."""
+    output = json_schema_to_guidance_output_template(cls.schema(), root=cls.schema())
+    return wrap_json_markdown(output)
+
+
+def json_schema_to_guidance_output_template(
+    schema: dict,
```

Comment:
Would it help if schema was a typed dict? Just a little confused about what the schema is supposed to look like here. Not sure if Scott's changes will clean this up though

Reply:
it's a json dict. I think that is as far as we can type it.

The schema is generated from pydantic model. looks something like this (generated with ChatGPT, but you get the idea):

```json
{
  "$schema": "http://json-schema.org/draft-07/schema#",
  "title": "NestedObject",
  "type": "object",
  "properties": {
    "title": {
      "type": "string"
    },
    "items": {
      "type": "array",
      "items": {
        "$ref": "#/definitions/Item"
      }
    }
  },
  "definitions": {
    "Item": {
      "type": "object",
      "properties": {
        "id": {
          "type": "integer"
        },
        "name": {
          "type": "string"
        },
        "sub_items": {
          "type": "array",
          "items": {
            "$ref": "#/definitions/SubItem"
          }
        }
      },
      "required": [
        "id",
        "name",
        "sub_items"
      ]
    },
    "SubItem": {
      "type": "object",
      "properties": {
        "name": {
          "type": "string"
        },
        "quantity": {
          "type": "integer"
        }
      },
      "required": [
        "name",
        "quantity"
      ]
    }
  },
  "required": [
    "title",
    "items"
  ]
}
```